### PR TITLE
Active Learning

### PR DESCRIPTION
**Pull request by jdridder**
This PR adds the feature of **_active learning_** meaning bayesian optimization with **_pure exploration_** to the repo. It includes a new predictive strategy named `ActiveLearningStrategy` that uses the acquisition function `[qNegIntegratedPosteriorVariance]` [link](https://botorch.org/api/_modules/botorch/acquisition/active_learning.html#qNegIntegratedPosteriorVariance) to choose new candidate points depending on how strongly they will reduce the variance of the posterior. Like proposed in this [issue](https://github.com/experimental-design/bofire/issues/331).

This new strategy can handle single and multiple output models. For multiple output models specific weights for each output variable can be supplied.

Furthermore, a tutorial notebook is provided that illustrates the usage of the `ActiveLearningStrategy` for single and multiple output cases. Please check the code for further adjustments. Hopefully, we can merge the PR soon.

### CODE REVIEW

File: bofire/data_models/strategies/predictives/active_learning.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+from bofire.data_models.objectives.api import Objective
+from bofire.data_models.strategies.predictives.botorch import BotorchStrategy
+
+
+class ActiveLearningStrategy(BotorchStrategy):
+    type: Literal["ActiveLearningStrategy"] = "ActiveLearningStrategy"
+    acquisition_function: AnyActiveLearningAcquisitionFunction = Field(
+        default_factory=lambda: qNegIntPosVar()
+    )
+    weights: Optional[Dict[str, float]] = Field(default_factory=lambda: None)
```
Comment (jduerholt) [inline bofire/data_models/strategies/predictives/active_learning.py:20]:
```suggestion
    weights: Optional[Dict[str, PositiveFloat]] = Field(default_factory=lambda: None)
```

Concerning the weights:
- you can import `PositiveFloat` from pydantic
- I think we concluded that we want to have the `weights` attribute under the acquisition function `qNegIntPostVar` as attribute as it is specific to this acquisition function and other active learning acquisition functions would not use this. So can you move it?
- We need then a `model_validator` with mode `"after"` in which we check that if `weights` are provided, the keys match the keys of the output features:
```python
@model_validator(mode="after")
def validate_acquisition_function(self):
   if isinstance(self.acquisition_function, qNegIntPosVar):
       if self.acquisition_function.weights is not None:
           if sorted(self.acquisition_function.weights.keys()) != sorted(self.domain.outputs.get_keys()):
               raise ValueError(balabala)
   return self
```

File: tests/bofire/data_models/specs/strategies.py [hunk 2]
```diff
@@ -105,6 +110,57 @@
...
+                ]
+            ),
+            outputs=Outputs(features=[ContinuousOutput(key="alpha")]),
+        ).model_dump(),
+        "acquisition_function": qNegIntPosVar(n_points=2048).model_dump(),
+        **strategy_commons,
+    },
+)
+
+specs.add_invalid(
```
Comment (jduerholt) [inline tests/bofire/data_models/specs/strategies.py:136]: This does not hold anymore as invalid, but you should modify it to test if the error is thrown if the keys of the weights does not match the output feature keys.

File: bofire/strategies/predictives/active_learning.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    def __init__(
+        self,
+        data_model: DataModel,
+        **kwargs,
+    ):
+        super().__init__(data_model=data_model, **kwargs)
+        self.acquisition_function = data_model.acquisition_function
+        self.weights = data_model.weights
+
+    def _get_acqfs(self, n) -> List[AcquisitionFunction]:
```
Comment (jduerholt) [inline bofire/strategies/predictives/active_learning.py:30]: Can you create a file `tests/bofire/strategies/predictives/test_active_learning.py` in which you test this method? And also check that the posterior transform etc. is correctly generated?

**Review by jduerholt — CHANGES_REQUESTED** [review]
Hi @jdridder,

thank you very much. I let a few comments. Should be easily done. There are also a few merge conflicts that has to be resolved.

Best,

Johannes

**Review by jdridder — COMMENTED** [review]
Thanks @jduerholt. I reviewed your suggestions and made the changes accordingly. Feel free to check for merging or to suggest further improvements. Also, please take a look at the conflicting files as I do not have write access to resolve the conflicts.

Regards,
JD

File: tests/bofire/strategies/test_active_learning.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        weight_list.append(aqcf_data_model.weights.get(key))
+        for key in benchmark.domain.outputs.get_keys()
+    ]
+    assert (
+        weight_list == acqf.posterior_transform.weights.tolist()
+    )  # Check whether the weights in the posterior_transfrom are set up correctly.
+    _ = recommender.ask(2)  # Check whether the optimization of the acqf works.
+
+
+def test_weight_keys_qNegIntPosVar():
```
Comment (jduerholt) [inline tests/bofire/strategies/test_active_learning.py:61]: Hmm, this test is not necessary, or? You are testing this already via the `add_invalid` ...

**Review by jduerholt — CHANGES_REQUESTED** [review]
Hi JD,

looks really good to me. Regarding the merge conflicts: you can just update your local main branch to the current upstream main, switch back to this development branch and do a `git merge main`, and resolve the conflicts. In case of problems, I can also do it for you. 

Best,

Johannes

File: tests/bofire/data_models/specs/strategies.py [hunk 5]
```diff
@@ -114,6 +119,62 @@
...
+            n_mc_samples=2048,
+            weights={
+                "alph_invalid": 0.1,
+                "beta_invalid": 0.9,
+            },
+        ).model_dump(),
+        **strategy_commons,
+    },
+    error=ValueError,
+    message="The provided keys for the weights must match the keys of the output features.",
```
Comment (jduerholt) [inline tests/bofire/data_models/specs/strategies.py:175]: This is the wrong error message, it should be: "The keys provided for the weights do not match the required keys of the output features." Have a look at `bofire/data_models/strategies/predictives/active_learning.py`

**Review by jduerholt — CHANGES_REQUESTED** [review]
Hey @jdridder,

I found one of the reasons why the tests are failing. See comment below. Just fix it and push again, then I will look for the second. 

Best,

Johannes

**Review by jduerholt — CHANGES_REQUESTED** [review]
Hi @jdridder,

it took me a while, but I found it. You have to register the `ActiveLearningStrategy` in `data_models/strategies/actual_strategy_type.py` as `ActualStrategy`.

Then it should work.

Best,

Johannes

**Review by jdridder — COMMENTED** [review]
Hi @jduerholt,

thanks for finding the error :) .
All missing changes are made. The last commits are stashed and forced-pushed into one final commit `[Bug fix tests ActiveLearning]`.
 Tests now run successfully.
Looking forward to merging.

Regards,
JD

**Review by jduerholt — APPROVED** [review]
Thank you very much @jdridder!

**Pull request merged**